File: C#/Circular Buffer/CircularBuffer.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class CircularBuffer<T>
{
    private List<T> _data;
    private int _writeIndex = 0;
    private int _readIndex = 0;

    public int Capacity { get; private set; } = 0;

    public int Count { get; private set; } = 0;

    private static int AdvanceIndex(int index, int length) => (index + 1) % length;

    public CircularBuffer(int capacity)
    {
        Capacity = capacity;
        Clear();
    }

    public void Write(T value)
    {
        if (this.Count == Capacity)
        {
            throw new InvalidOperationException("Buffer full");
        }

        _data[_writeIndex] = value;
        _writeIndex = AdvanceIndex(_writeIndex, Capacity);
        Count++;
    }

    public T Read()
    {
        if (Count == 0)
        {
            throw new InvalidOperationException("Buffer Empty");
        }

        T value = this._data[_readIndex];

        _data[_readIndex] = default; // Reset to default value (null for reference types)
        _readIndex = AdvanceIndex(_readIndex, Capacity);

        Count--;

        return value;
    }

    public void Overwrite(T value)
    {
        if (Count != Capacity)
        {
            Write(value);
        }
        else
        {
            _data[_readIndex] = value;
            _readIndex = AdvanceIndex(_readIndex, Capacity);
        }
    }

    public void Clear()
    {
        _data = new List<T>(Capacity);

        for (int i = 0; i < Capacity; i++)
        {
            _data.Add(default);
        }

        this.Count = 0;
        this._writeIndex = 0;
        this._readIndex = 0;
    }
}

File: TypeScript/circular-buffer.ts
const advanceIndex = (index: number, length: number): number => {
  return (index + 1) % length; 
};

interface Indicies {
  write: number;
  read: number;
}

export class CircularBuffer<T> {
  private data: (T|undefined)[];
  private usedSpaces: number;
  private index: Indicies;

  constructor(length: number) {
    this.data = new Array(length).fill(undefined);
    this.usedSpaces = 0;
    this.index = { write:0, read:0 };
  }

  get count(): number {
    return this.usedSpaces;
  }

  write(data: T): void{
    if(this.usedSpaces === this.data.length) {
      throw new BufferFullError();
    }

    this.data[this.index.write] = data;
    this.index.write = advanceIndex(this.index.write, this.data.length);
    this.usedSpaces++;
  }

  read(): (T|undefined) {
    if(this.usedSpaces === 0) {
      throw new BufferEmptyError();
    }

    let value = this.data[this.index.read];

    if(value != undefined) {
      this.data[this.index.read] = undefined;
      this.usedSpaces--;
    }

    this.index.read = advanceIndex(this.index.read, this.data.length);

    return value;
  }

  forceWrite(data: T): void {
    if(this.data[this.index.write] === undefined) {
      this.write(data);
    }
    else {
      this.data[this.index.read] = data;
      this.index.read = advanceIndex(this.index.read, this.data.length);
    }
  }

  clear(): void {
    this.data = new Array(this.data.length).fill(undefined);
    this.usedSpaces = 0;
    this.index = { write:0, read:0 };
  }
}
 
export class BufferFullError extends Error {
  constructor() {
    super("Can't write to a full buffer");
  }
}
  
export class BufferEmptyError extends Error {
  constructor() {
    super("Can't read from an empty buffer");
  }
}

const buffer = new CircularBuffer<string>(3);
buffer.write('1');
buffer.write('2');
buffer.write('3');

console.log(buffer.read()) // '1';

buffer.write('4');
buffer.forceWrite('5');

console.log(buffer.read()) // '3';
console.log(buffer.read()) // '4';
console.log(buffer.read()) // '5';

console.log(buffer.count) // '0;'